crates/ruma-events/src/secret/request.rs

```diff
@@ -0,0 +1,165 @@
+    #[serde(flatten)]
+    pub action: RequestAction,
+
+    /// The ID of the device requesting the event.
+    pub requesting_device_id: DeviceIdBox,
+
+    /// A random string uniquely identifying (with respect to the requester and the target) the
+    /// target for a secret.
+    ///
+    /// If the secret is requested from multiple devices at the same time, the same ID may be used
+    /// for every target. The same ID is also used in order to cancel a previous request.
+    pub request_id: String,
+}
+
+impl SecretRequestEventContent {
+    /// Creates a new `SecretRequestEventContent` with the given action, requesting device ID and
+    /// request ID.
+    pub fn new(
+        action: RequestAction,
+        requesting_device_id: DeviceIdBox,
+        request_id: String,
+    ) -> Self {
+        Self { action, requesting_device_id, request_id }
+    }
+}
+
+/// Action for a *m.secret.request* event.
+#[derive(Clone, Debug, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub enum RequestAction {
+    /// Request a secret by its name.
+    Request(SecretName),
+
+    /// Cancel a request for a secret.
+    RequestCancellation,
+
+    #[doc(hidden)]
+    _Custom(String),
+}
+
+impl Serialize for RequestAction {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        match self {
+            Self::Request(name) => {
+                let mut st = serializer.serialize_map(Some(2))?;
+                st.serialize_entry("name", name)?;
+                st.serialize_entry("action", "request")?;
+                st.end()
+            }
+            Self::RequestCancellation => {
+                let mut st = serializer.serialize_map(Some(1))?;
+                st.serialize_entry("action", "request_cancellation")?;
+                st.end()
+            }
+            RequestAction::_Custom(custom) => serializer.serialize_str(custom),
+        }
+    }
+}
+
+/// The name of a secret.
+#[derive(Clone, Debug, StringEnum)]
+pub enum SecretName {
+    /// Cross-signing master key (m.cross_signing.master).
+    #[ruma_enum(rename = "m.cross_signing.master")]
+    CrossSigningMasterKey,
+
+    /// Cross-signing user-signing key (m.cross_signing.user_signing).
+    #[ruma_enum(rename = "m.cross_signing.user_signing")]
+    CrossSigningUserSigningKey,
+
+    /// Cross-signing self-signing key (m.cross_signing.self_signing).
+    #[ruma_enum(rename = "m.cross_signing.self_signing")]
+    CrossSigningSelfSigningKey,
+
+    /// Recovery key (m.megolm_backup.v1.recovery_key).
+    #[ruma_enum(rename = "m.megolm_backup.v1.recovery_key")]
+    RecoveryKey,
+
+    /// Custom secret name.
+    Custom(String),
+}
+
+#[cfg(test)]
+mod test {
+    use super::{RequestAction, SecretName, SecretRequestEventContent};
+    use serde_json::{json, to_value as to_json_value};
+
+    #[test]
+    fn secret_request_serialization() {
```

Comment:
Can we test deserialization as well? It's not uncommon for only one of them to be broken.

Reply:
I've added tests, and it turns out the derived `Deserialization` implementation didn't work.
I wasn't sure about what to do with custom `action` variants when deserializing, so I made it discard the `name` field the same way it does when the `action` is `"request_cancellation"`.

Reply:
I think discarding the secret name for custom actions makes sense, and it's forwards compatible with a richer API for custom actions later.